### ISSUE TITLE
Possible: Update Matlab inteface to use a slightly more structured interface

### ISSUE THREAD

**Issue by msevestre** [Open-Systems-Pharmacology/PK-Sim#483]
@Yuri05 TO be discussed. I'd like to know what is supported by matlab and see what it would mean for the Matlab code. Hopefully access to PKSim.Matlab.dll is encapsulated in one or two files so that changing the PKSim interface would not break user defined scripts

**Comment by StephanSchaller** [comment]
How about the R interface?

**Comment by msevestre** [comment]
@StephanSchaller what do you mean?I am not aware of any interface between PK-Sim and R.

**Comment by StephanSchaller** [comment]
@msevestre, from the limited details you give I am not entirely sure what it is you are pointing out. I don't know what you are changing in the PKSim interface, but if it may have impact on MATLAB scripts, it may also impact R scripts, right? And by R interface i meant PKSim/MoBi interface with R (the .dll library in "..\libs\i386" ?).

**Comment by msevestre** [comment]
> it may also impact R scripts, right?
I don't think so. PKSim provides a component `PKSim.Matlab.dll` that is loaded in Matlab and allows the user to create population and individual directly from matlab code. (e.g. get arrays of age dependent parameters for examples) using this [script](https://github.com/Open-Systems-Pharmacology/Matlab-Toolbox-Core/blob/8a18f57abeef17a1cc14aa638aeb8512d2a2ebcb/code/auxiliaries/PKSimCreateIndividual.m). There is nothing like that in R. 

Because the PKSim.Matlab.dll is going to change, the script will need to be adjusted.

**Comment by StephanSchaller** [comment]
Okay, I get you...

> There is nothing like that in R. <

Well, that is a petty ;-) How could we change that?

**Comment by Yuri05** [comment]
>>How could we change that?

[Contribute](https://github.com/Open-Systems-Pharmacology/PK-Sim/issues/483#issuecomment-349662493) ;-)

`PKSim.Matlab.dll` is just a .NET component, which can be called apart from PK-Sim e.g. to create individuals or populations. Nothing Matlab-specific in fact (@msevestre correct me if I'm wrong)

A am not aware about the options for calling .NET from R, but there will be some for sure. So if somebody will implement the bridge to `PKSim.Matlab.dll` in R - we will have the same functionality there.

**Comment by StephanSchaller** [comment]
We would be happy to, but lacking resources... can you help kick-start @PavelBal to do so?

**Comment by PavelBal** [comment]
It seems like R can load "normal" (C/C++) DLL but not .NET DLL. The proposed workaround is to generate a COM DLL of the assembly (in our case the PKSim.Matlab.dll) and a C++ DLL that calls the COM, and load the C++ DLL in R:

https://stackoverflow.com/questions/17214741/is-it-possible-to-call-a-r-statistics-function-to-optimize-c-sharp-function

Sounds weird, but should be possible?

**Comment by Yuri05** [comment]
Check this out: 
* https://cran.r-project.org/web/packages/rDotNet/index.html or 
* https://rclr.codeplex.com/

**Comment by msevestre** [comment]
I would not go the COM route. This opens a box of pandora that will result in pain :). We finally managed to get rid of COM alltogether in PKSim 5.0 and I do not want to see that ever again

**Comment by msevestre** [comment]
@all: My suggestion would be to move this discussion somewhere else. This issue is a Task that I need to accomplish for one of the current development and this will be closed and (hidden/lost) once finished. 
@StephanSchaller Suggestion is to create an entry in the forum and gauging if people are interested in calling CreateIndividual and CreatePopulation from R directly as it is done in Matlab. 
Or open at least a new issue in PKSim :) Thanks

**Comment by msevestre** [comment]
Interface updated to use Snapshot instead of specific objects.

**Comment by StephanSchaller** [comment]
will do (no need to gauge ;-))
